File: Custom_Comparators_&_Coordinate_Compression/Silver/P1_2018-LifeGuards/P1_2018-LifeGuards.py
# Sorting/Sweep Line/greedy/intervals method
# T: O(n log n), O(n), where n is amount of shifts

class Shift_Event:
	def __init__(self, time = int, lifeguard_id = int, is_start = bool) -> None:
		self.time = time
		self.lifeguard_id = lifeguard_id
		self.is_start = is_start

# Open input file
with open('lifeguards.in', 'r') as f:
	# First line: n = num of cow lifeguards
	n = int(f.readline())
	# For the next n lines assign start/end shifts with lifeguard ids
	shifts = []
	for lg_id in range(n):
		start, end = map(int, f.readline().split())
		shifts.append(Shift_Event(start, lg_id, True))
		shifts.append(Shift_Event(end, lg_id, False))

# Sort by time only
shifts.sort(key=lambda se: se.time)

total_time = 0
on_duty = set()			# tracked by lifeguard id
alone_time = [0] * n	# idx is lifeguard id, value is total alone time
prev_time = 0
# iterate through shift events calculating result using sweep line approach
for se in shifts:
	if len(on_duty) == 1:	# if current lifeguard is alone
		alone_time[next(iter(on_duty))] += se.time - prev_time
	if len(on_duty) > 0:
		total_time += se.time - prev_time
	if se.is_start:
		on_duty.add(se.lifeguard_id)
	else:
		on_duty.remove(se.lifeguard_id)
	prev_time = se.time
# Write total time - life guard with least alone time to output file
print(total_time - min(alone_time), file=open('lifeguards.out', 'w'))

File: Custom_Comparators_and_Coordinate_Compression/Silver/P1_2018-LifeGuards/P1_2018-LifeGuards.cpp
// Custom Comparators + Sorting + Sweep Line Approach
// T: O(n log n), M: O(n), where n is amount of shifts

// #include <bits/stdc++.h>

#include <iostream>
#include <vector>
#include <algorithm>
#include <unordered_set>

using namespace std;

/*
class ShiftEvent {
public:
	int time, lifeguardID;
	bool isStart;
	ShiftEvent(int time, int lifeguardID, bool isStart) {
		this->time = time;
		this->lifeguardID = lifeguardID;
		this->isStart = isStart;
	}
};
*/

struct ShiftEvent {
	int time, lifeguardID;
	bool isStart;
};

int main() {
	// Open input file
	freopen("lifeguards.in", "r", stdin);
	// First line: n = number of cow lifeguards
	int n;
	cin >> n;
	// For the next n lines assign start/end shifts with lifeguard IDs (ID is based on idx)
	vector<ShiftEvent> shifts;
	for (int lg_id = 0; lg_id < n; lg_id++) {
		int start, end;
		cin >> start >> end;
		shifts.push_back({start, lg_id, true});
		shifts.push_back({end, lg_id, false});
	}

	// Sort the shifts by time only
	sort(shifts.begin(), shifts.end(), [](ShiftEvent a, ShiftEvent b) {return a.time < b.time;});

	int aloneTime[n] = {};				// idx is lifeguard id, value is total alone time
	unordered_set<int> onDuty;			// tracked by lifeguard id
	int prevTime = 0, totalTime = 0;
	// iterate through shift events calculating result using sweep line approach
	for (ShiftEvent &se : shifts) {
		if (onDuty.size() == 1)
			aloneTime[*onDuty.begin()] += se.time - prevTime;
		
		if (onDuty.size() > 0)
			totalTime += se.time - prevTime;
		
		if (se.isStart)
			onDuty.insert(se.lifeguardID);
		else
			onDuty.erase(se.lifeguardID);

		prevTime = se.time;
	}

	// Write total time - life guard with least alone time to output file
	freopen("lifeguards.out", "w", stdout);
	cout << totalTime - *min_element(aloneTime, aloneTime + n) << endl;
}